src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Threading;
+
+namespace Microsoft.JSInterop.Infrastructure
+{
+    internal sealed class ByteArrayJsonConverter : JsonConverter<byte[]>
+    {
+        // Atomically incrementing unique identifier for serializing byte arrays.
+        private static long _byteArrayId;
+
+        internal static readonly JsonEncodedText ByteArrayRefKey = JsonEncodedText.Encode("__byte[]");
+
+        public ByteArrayJsonConverter(JSRuntime jSRuntime)
+        {
+            JSRuntime = jSRuntime;
+        }
+
+        public JSRuntime JSRuntime { get; }
+
+        public override bool CanConvert(Type typeToConvert) => typeToConvert == typeof(byte[]);
+
+        public override byte[]? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (JSRuntime.ByteArraysToBeRevived.Count == 0)
+            {
+                throw new JsonException("ByteArraysToBeRevived is empty.");
+            }
+
+            long? byteArrayIndex = null;
+
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
+            {
+                if (reader.TokenType == JsonTokenType.PropertyName)
+                {
+                    if (byteArrayIndex is null && reader.ValueTextEquals(ByteArrayRefKey.EncodedUtf8Bytes))
+                    {
+                        reader.Read();
+                        byteArrayIndex = reader.GetInt64();
+                    }
+                    else
+                    {
+                        throw new JsonException($"Unexpected JSON property {reader.GetString()}.");
+                    }
+                }
+                else
+                {
+                    throw new JsonException($"Unexpected JSON Token {reader.TokenType}.");
+                }
+            }
+
+            if (byteArrayIndex is null)
+            {
+                throw new JsonException($"Required property {ByteArrayRefKey} not found.");
+            }
+
+            if (byteArrayIndex >= JSRuntime.ByteArraysToBeRevived.Count || byteArrayIndex < 0)
+            {
+                throw new JsonException($"Byte array {byteArrayIndex} not found.");
+            }
+
+            var byteArray = JSRuntime.ByteArraysToBeRevived.Buffer[byteArrayIndex.Value];
+            if (byteArray is null || byteArray.Length == 0)
+            {
+                return Array.Empty<byte>();
+            }
```

Comment:
Could you clarify what this block is for? Why would one of the array entries be `null`, and why do we care about normalizing already-allocated empty array instances to be `Array.Empty<byte>()`?

Reply:
In the case where [message pack fails to read the byte arrays](https://github.com/dotnet/aspnetcore/pull/33015/files#diff-8bdd71729dca8e301353f4156cdf02e587c99ba34e98d23d8361641428e53130R43) (potentially because the user provided a null byte array).

> why do we care about normalizing already-allocated empty array instances to be Array.Empty<byte>()?

We don't/shouldn't. Thanks for pointing that out. Removed.

Reply:
Given the thing you pointed out below, removing this block entirely.